### PR TITLE
feat(protocol-designer): edit instruments modal

### PR DESCRIPTION
**Pull request by jerader**
closes AUTH-661

# Overview

Make the edit instruments modal and wire it up, there are some TODOs stilllllll, currently blocked by designs 🫠 :
1. https://opentrons.atlassian.net/browse/AUTH-768

## Test Plan and Hands on Testing

Create a flex or ot-2 protocol and edit the pipettes and tipracks. the tiprack entities on the deck map should also delete/regenerate

Also the flex should allow to add/remove the gripper

## Changelog

- create edit instruments modal
- wire up to the protocol overview page

## Risk assessment

low

### CODE REVIEW

File: protocol-designer/src/organisms/EditInstrumentsModal/CreatePipettes.ts [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+import type { StepIdType } from '../../form-types'
+
+const adapter96ChannelDefUri = 'opentrons/opentrons_flex_96_tiprack_adapter/1'
+
+type PipetteFieldsData = Omit<
+  PipetteOnDeck,
+  'id' | 'spec' | 'tiprackLabwareDef'
+>
+
+export const createPipettes = (
```
Comment (jerader) [inline protocol-designer/src/organisms/EditInstrumentsModal/CreatePipettes.ts]: this fn is big! it is basically a copy of legacy code though, so it should work as expected. Also includes the ordered step ids to replace the pipette if you've udpated it and its used in a step.

File: protocol-designer/src/organisms/EditInstrumentsModal/CreatePipettes.ts
Comment (koji) [inline protocol-designer/src/organisms/EditInstrumentsModal/CreatePipettes.ts:1]: nit

`createPipettes.ts`

File: protocol-designer/src/organisms/EditInstrumentsModal/CreatePipettes.ts [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+        (
+          p: typeof nextPipettes[keyof typeof nextPipettes]
+        ): NormalizedPipette => ({
+          id: p.id,
+          name: p.name,
+          tiprackDefURI: p.tiprackDefURI,
+        })
+      )
+    )
+  )
```
Comment (koji) [inline protocol-designer/src/organisms/EditInstrumentsModal/CreatePipettes.ts]: nit

```suggestion
  dispatch(
    stepFormActions.createPipettes(
      mapValues(
        nextPipettes,
        (
          {id, name, tiprackDefURI}: typeof nextPipettes[keyof typeof nextPipettes]
        ): NormalizedPipette => ({
          id,
          name,
          tiprackDefURI,
        })
      )
    )
  )
```

File: protocol-designer/src/organisms/EditInstrumentsModal/CreatePipettes.ts [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+      return replacementId && replacementId !== -1
+        ? { ...acc, [deletedId]: replacementId }
+        : acc
+    },
+    {}
+  )
+
+  const pipettesWithNewTipracks: string[] = filter(
+    nextPipettes,
+    (nextPipette: typeof nextPipettes[keyof typeof nextPipettes]) => {
```
Comment (koji) [inline protocol-designer/src/organisms/EditInstrumentsModal/CreatePipettes.ts]:
```suggestion
    ({ id: newPipetteId, tiprackDefURI: nextTips }: typeof nextPipettes[keyof typeof nextPipettes]) => {
```

File: protocol-designer/src/organisms/EditInstrumentsModal/CreatePipettes.ts [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+  }
+
+  // substitute deleted pipettes with new pipettes on the same mount, if any
+  if (!isEmpty(substitutionMap) && orderedStepIds.length > 0) {
+    // NOTE: using start/end here is meant to future-proof this action for multi-step editing
+    dispatch(
+      stepFormActions.substituteStepFormPipettes({
+        substitutionMap,
+        startStepId: orderedStepIds[0],
+        // @ts-expect-error(sa, 2021-6-22): last might return undefined
```
Comment (koji) [inline protocol-designer/src/organisms/EditInstrumentsModal/CreatePipettes.ts]: still need this?

File: protocol-designer/src/organisms/EditInstrumentsModal/CreatePipettes.ts [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+  pipettes: { [pipetteId: string]: PipetteOnDeck },
+  orderedStepIds: StepIdType[],
+  dispatch: ThunkDispatch<any>,
+  mount: PipetteMount,
+  selectedPip: PipetteName,
+  selectedTips: string[],
+  leftPip?: PipetteOnDeck,
+  rightPip?: PipetteOnDeck
+): void => {
+  const otherPip = mount === 'left' ? rightPip : leftPip
```
Comment (koji) [inline protocol-designer/src/organisms/EditInstrumentsModal/CreatePipettes.ts]: nit

```suggestion
  const oppositePipette = mount === 'left' ? rightPip : leftPip
```

File: protocol-designer/src/pages/ProtocolOverview/index.tsx [hunk 6]
```diff
@@ -320,7 +330,7 @@ export function ProtocolOverview(): JSX.Element {
                     <ListItemDescriptor
                       type="default"
                       description={t('extension')}
-                      content={gripper != null ? t(`$gripper.name}`) : 'N/A'}
+                      content={isGripperAttached ? t('gripper') : 'N/A'}
```
Comment (koji) [inline protocol-designer/src/pages/ProtocolOverview/index.tsx]:
```suggestion
                      content={isGripperAttached ? t('gripper') : t('na')}
```

**Review by ncdiehl11 — COMMENTED** [review]
Functionality looks good, but I am getting a whitescreen when editing a pipette. Looks like this arises from being able to save without inputting generation, tips, etc.
Screenshot 2024-09-05 at 11 39 36 AM

**Review by ncdiehl11 — APPROVED** [review]
lookin good!

**Pull request merged**